geode-core/src/main/java/org/apache/geode/cache/client/internal/PoolImpl.java

```diff
@@ -1581,4 +1582,27 @@ public int getPendingEventCount() {
   public int getSubscriptionTimeoutMultiplier() {
     return subscriptionTimeoutMultiplier;
   }
+
+  public int calculateRetryAttempts(Throwable cause) {
+
+    int maxRetryAttempts = getRetryAttempts();
+
+    if (maxRetryAttempts == PoolFactory.DEFAULT_RETRY_ATTEMPTS) {
+      // If the retryAttempt is set to default(-1). Try executing on all servers once.
+      // As calculating number of servers involves sending message to locator, it is
+      // done only when there is an exception.
+      if (cause instanceof ServerConnectivityException
+          && cause.getMessage().contains(ConnectionManagerImpl.BORROW_CONN_ERROR_MSG)) {
+        // The client was unable to establish a connection before sending the
+        // request.
+        maxRetryAttempts = getConnectionSource().getAllServers().size();
+      } else {
+        // The request was sent once.
+        maxRetryAttempts = getConnectionSource().getAllServers().size() - 1;
```

Comment:
Do we know that in all other cases we want -1.  Would it be safer to do size() for all cases and isolate the specific scenarios where we would want to do a -1?

Reply:
There's a discussion on that. We can use size() -1 as default SCE, and identify each special case to not to decrease 1. Another alternative is: use size() as default SCE, only to use size()-1 for timeout error msg.